sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -785,17 +786,19 @@ case class Cast(
       buildCast[Boolean](_, b => if (b) 1 else 0)
     case DateType =>
       buildCast[Int](_, d => null)
-    case TimestampType if ansiEnabled =>
+    case TimestampType =>
       buildCast[Long](_, t => {
         val longValue = timestampToLong(t)
         if (longValue == longValue.toInt) {
           longValue.toInt
         } else {
-          throw QueryExecutionErrors.castingCauseOverflowError(t, from, IntegerType)
+          if (ansiEnabled) {
+            throw QueryExecutionErrors.castingCauseOverflowError(t, from, IntegerType)
+          } else {
+            null
+          }
```

Comment:
While I'm working on related stuff, I take a look how Spark handles casting timestamp to integer. If no special reason, this behavior doesn't look correct to me and is not common way for overflowing when casting other types under non-ansi mode.

Reply:
Let me know if I miss anything here.

Reply:
Do you know when the current behavior was introduced, @viirya ?

Reply:
As far as I can check, it is current behavior at the beginning. So that's why I wonder if I miss anything. But indeed, it might be rare to cast timestamp to byte/short/int. Maybe it is why this is not caught? 🤔

Reply:
Got it. Thank you for checking.

Reply:
I checked Postgres.
`select CAST('32' as int);` output 32.
`select CAST(TIMESTAMP '1900-05-05 18:34:56.1' as int);` output `SQL error [42846]: ERROR: cannot cast type timestamp without time zone to integer`
`select CAST(TIMESTAMP WITH TIME ZONE '1900-05-05 18:34:56.1' as int);` output `SQL error [42846]: ERROR: cannot cast type timestamp with time zone to integer`.

Reply:
It seems we should throw error even if without ansi mode.

Reply:
Oracle.
`SELECT CAST('2' AS int) FROM DUAL;` output 2.
SELECT CAST(TO_TIMESTAMP('1900-05-05 18:34:56.10', 'YYYY-MM-DD HH24:MI:SS.FF') AS int) FROM DUAL; output `SQL 错误 [932] [42000]: ORA-00932: 数据类型不一致: 应为 NUMBER, 但却获得 TIMESTAMP`